sdk/search/search-documents/src/searchIndexingBufferedSender.ts

```diff
@@ -377,9 +386,35 @@ export class SearchIndexingBufferedSender<T> {
       this.batchObject = new IndexDocumentsBatch<T>();
       while (actions.length > 0) {
         const actionsToSend = actions.splice(0, this.initialBatchActionCount);
-        await this.submitDocuments(actionsToSend, options);
+        await this.submitDocuments(this.pruneActions(actionsToSend, actions), options);
+      }
+    }
+  }
+
+  private pruneActions(
+    batch: IndexDocumentsAction<T>[],
+    actions: IndexDocumentsAction<T>[]
+  ): IndexDocumentsAction<T>[] {
+    let counter: number = 0;
+    const hashSet: Set<string> = new Set<string>();
+    const resultBatch: IndexDocumentsAction<T>[] = [];
+
+    while (counter < batch.length) {
+      const { __actionType, ...document } = batch[counter];
+      // @ts-ignore
```

Comment:
why do we have ts-ignore here?

Reply:
If I do not add this here, I get the following error:

```
src/searchIndexingBufferedSender.ts:405:45 - error TS2345: Argument of type 'IndexDocumentsAction<T>' is not assignable to parameter of type 'T'.
  'T' could be instantiated with an arbitrary type which could be unrelated to 'IndexDocumentsAction<T>'.

405       const key = this.documentKeyRetriever(document);
```

Reply:
This is interesting. Since merge and delete operations don't need to specify the entire document, IndexDocumentsAction<T> maps to a Partial<T> plus the action property.

Since the action can be a partial document, that means the mapping function might not be getting the whole document. This should still be fine as the partial document **must** contain the key, but TS doesn't know that, so what it wants is for documentKeyRetriever to look like this:

```ts
documentKeyRetriever: (document: T | IndexDocumentsAction<T>) => string
```

I notice that buffered sender doesn't actually ever take in partial documents, so I think in this case it's safe to avoid it with

```ts
const key = this.documentKeyRetriever(document as unknown as T);
```

Which is still much better than `ts-ignore` as `ts-ignore` disables all checking instead of just one type check.